csvcubed/csvcubed/readers/columnschema.py

```diff
@@ -0,0 +1,447 @@
+"""
+Models
+------
+
+config.json V1.0 column mapping models.
+"""
+from dataclasses import dataclass
+from typing import List, Union, Optional, TypeVar
+from pathlib import Path
+
+import uritemplate
+from csvcubed.models.cube import (
+    ExistingQbAttributeLiteral,
+    NewQbAttributeLiteral,
+    CompositeQbCodeList,
+    CatalogMetadata,
+    DuplicatedQbConcept,
+    NewQbCodeList,
+)
+from csvcubedmodels.dataclassbase import DataClassBase
+
+from csvcubed.inputs import pandas_input_to_columnar_optional_str
+from csvcubed.models.cube.qb.components import (
+    NewQbDimension,
+    ExistingQbDimension,
+    NewQbAttribute,
+    NewQbAttributeValue,
+    ExistingQbAttribute,
+    NewQbUnit,
+    ExistingQbUnit,
+    QbMultiUnits,
+    QbMultiMeasureDimension,
+    QbMultiMeasureObservationValue,
+    QbSingleMeasureObservationValue,
+    ExistingQbCodeList,
+    ExistingQbMeasure,
+    NewQbMeasure,
+    QbObservationValue,
+    NewQbCodeListInCsvW,
+)
+from csvcubed.inputs import PandasDataTypes
+from csvcubed.utils.uri import looks_like_uri, csvw_column_name_safe
+
+T = TypeVar("T", bound=object)
+
+
+@dataclass
+class SchemaBaseClass(DataClassBase):
+    ...
+
+
+@dataclass
+class Dimension(SchemaBaseClass):
+    # Schema property - but removed for json to dataclass mapping
+    # type: # str = "dimension"
+
+    # Properties only available for New Dimension
+    label: Optional[str] = None
+    description: Optional[str] = None
+    definition_uri: Optional[str] = None
+
+    # Properties common to both New and Existing Dimension
+    from_existing: Optional[str] = None
+    cell_uri_template: Optional[str] = None
+    code_list: Optional[Union[str, bool]] = False
+
+    def map_to_qb_dimension(
+        self, label: str, data: PandasDataTypes, json_parent_dir: Path
+    ) -> Union[NewQbDimension, ExistingQbDimension]:
+
+        # No label, description or code_list means its likely to be an ExistingDimension
+        if not any([self.label, self.description, self.code_list]):
+            # but requires a valid from_existing uri str
+            if isinstance(self.from_existing, str) :
+                return ExistingQbDimension(self.from_existing)
+            else:
+                raise ValueError(
+                    "Existing Dimension - from_existing property was not a valid URI"
+                )
+
+        # label, description or code_list present therefore it will be a NewDimension
+        else:
+            new_dimension = NewQbDimension.from_data(
+                label=label,
+                data=data,
+                description=self.description,
+                parent_dimension_uri=self.from_existing,
+                source_uri=self.definition_uri,
+            )
+            # new_dimension.code_list = self._get_code_list(new_dimension, json_parent_dir)
```

Comment:
We need this to be uncommented.

Reply:
Done, with _get_code_list method updated.